src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -948,38 +964,48 @@ internal bool VerifyRemoteCertificate(RemoteCertValidationCallback? remoteCertVa
                     sslPolicyErrors |= CertificateValidationPal.VerifyCertificateProperties(
                         _securityContext!,
                         chain,
-                        remoteCertificateEx,
+                        _remoteCertificate,
                         _sslAuthenticationOptions.CheckCertName,
                         _sslAuthenticationOptions.IsServer,
                         _sslAuthenticationOptions.TargetHost);
                 }
 
                 if (remoteCertValidationCallback != null)
                 {
-                    success = remoteCertValidationCallback(_sslAuthenticationOptions.TargetHost, remoteCertificateEx, chain, sslPolicyErrors);
+                    object? sender = _ssl;
+                    if (sender == null)
+                    {
+                        throw new ObjectDisposedException(nameof(SslStream));
+                    }
+
+                    success = remoteCertValidationCallback(sender, _remoteCertificate, chain, sslPolicyErrors);
                 }
                 else
                 {
-                    if (sslPolicyErrors == SslPolicyErrors.RemoteCertificateNotAvailable && !_sslAuthenticationOptions.RemoteCertRequired)
-                    {
-                        success = true;
-                    }
-                    else
+                    if (!RemoteCertRequired)
                     {
-                        success = (sslPolicyErrors == SslPolicyErrors.None);
+                        sslPolicyErrors &= ~SslPolicyErrors.RemoteCertificateNotAvailable;
                     }
+
+                    success = (sslPolicyErrors == SslPolicyErrors.None);
                 }
 
                 if (NetEventSource.Log.IsEnabled())
                 {
                     LogCertificateValidation(remoteCertValidationCallback, sslPolicyErrors, success, chain!);
-                    if (NetEventSource.Log.IsEnabled())
-                        NetEventSource.Info(this, $"Cert validation, remote cert = {remoteCertificateEx}");
+                    NetEventSource.Info(this, $"Cert validation, remote cert = {_remoteCertificate}");
                 }
 
                 if (!success)
                 {
                     alertToken = CreateFatalHandshakeAlertToken(sslPolicyErrors, chain!);
+                    if (chain != null)
```

Comment:
The line above this does `chain~`, so doesn't that mean chain can't be null here?

Reply:
The chain would not be null only is we get remote certificate. For example, we get to this path as well in cases when server asked for client certificate but client dod not provide any.